Fix: CapRover Dashboard App: Details of Workers disappear after few seconds

Mahmoud-Emad opened the pull request:
### Description

Fix listing Caprover workers

### Changes

- Send both of the deployments, leader, and workers.
- Call the 'getGrafanaUrl' on mount the component.
- Delete the full cluster.
- Ensure that workers are correctly removed from the deployment list when the leader is deleted to maintain data integrity.
- Merged the latest changes from the development branch to keep the codebase up to date.
- Conducted thorough testing to verify the complete functionality and ensure no regressions were introduced.

### Test Scenarios

1. Deploy `leader`, and `worker`, Go go to the contracts page and delete the deployed `leader`.
2. Deploy `leader`, attach a `worker` to it then delete the whole cluster just in time without refreshing the page.
3. Deploy `leader`, and `worker`, delete the deployed `worker` and attach another, then delete the whole cluster just in time without refreshing the page.
4. Deploy a cluster with 1 leader and 2 workers check the deployment listing dialog and make sure all of them are listed
3. Delete a cluster with the leader and more than 3 workers and make sure all of them are deleted.

PS: Make sure no errors on the Caprover listing page.

### Related Issues

- https://github.com/threefoldtech/tfgrid-sdk-ts/issues/3281

### Screenshots/Videos

[Screencast from 09-03-2024 05:27:14 PM.webm](https://github.com/user-attachments/assets/45752f9d-d688-4869-8327-f68c5aad1c65)

![image](https://github.com/user-attachments/assets/51593928-7848-4744-b01d-f7d4474a3c57)

### Checklist

- [ ] Tests included
- [ ] Build pass
- [ ] Documentation
- [ ] Code format and docstrings
- [ ] Screenshots/Video attached (needed for UI changes)

Mahmoud-Emad committed:
chore: Build the grafana URL on mount.

Mahmoud-Emad committed:
Merge branch 'development' of https://github.com/threefoldtech/tfgrid-sdk-ts into development_caprover

Mahmoud-Emad committed:
fix: Fix an issue with loading Grafana URL

Mahmoud-Emad committed:
fix: Fix the listing issue:

- Send the both of the deployments, leader and workers.
- Call the 'getGrafanaUrl' on mounting the component.

Mahmoud-Emad committed:
feat: Enhance Deletion of Caprover Cluster:

- Enhanced Deletion Dialog: displays all attached workers in the deletion dialog to prompt user confirmation before proceeding.
- Worker Deletion Logic: Iterates through the machine's interface, extracts the associated worker nodes, and deletes them efficiently.
- Refactored Delete Function: Improved the delete function.

MohamedElmdary commented on packages/playground/src/weblets/tf_deployment_list.vue:
```diff
@@ -464,53 +474,114 @@ const deletingDialog = ref(false);
 const table = ref() as Ref<{ loadDeployments(): void }>;
 const gridStore = useGrid();
 const grid = gridStore.client as GridClient;
+const hasWorkers = computed(
+  () => selectedItems.value.map(item => item.projectName.includes(ProjectName.Caprover.toLocaleLowerCase())).length > 0,
+);
```
I think you mean filter instead of map?

Mahmoud-Emad committed:
fix: Use `map.some()` instead of map.length.

- Used the some() method since it'll be faster than the `filter` method.
- Update the checker to check if the selected items have workers.

0oM4R reviewed (commented):
I was able to list and delete all workers fine, but on adding new one i got this error
![Screenshot from 2024-09-12 12-06-51](https://github.com/user-attachments/assets/7e74f8d4-e825-42c7-ba55-fedb3129b678)

Mahmoud-Emad committed:
Merge branch 'development' into development_caprover

Mahmoud-Emad commented:
> I was able to list and delete all workers fine, but on adding new one i got this error ![Screenshot from 2024-09-12 12-06-51](https://private-user-images.githubusercontent.com/62248851/366802653-7e74f8d4-e825-42c7-ba55-fedb3129b678.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjY0Njk3NDksIm5iZiI6MTcyNjQ2OTQ0OSwicGF0aCI6Ii82MjI0ODg1MS8zNjY4MDI2NTMtN2U3NGY4ZDQtZTgyNS00MmM3LWJhNTUtZmVkYjMxMjliNjc4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTE2VDA2NTA0OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU3OWU2ODJmOGJhMmYyYjQxOGNlNDQ1YWZkMjUxNDEwYjZiNGNjNTZmNjQzNTdlNTE5NTA0NjhjMzZkZjBlM2QmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.XDjmyKOJlV58hXgyQFVveIbSgmPPEtUMNr7TJs-TLic)

This happened because the worker is listed instead of the master, which is wrong

Mahmoud-Emad committed:
Merge branch 'development' into development_caprover

Mahmoud-Emad committed:
fix: Fix listing Caprover:

- Push the master into the first index of the result array
- Push the workers into the master, second index of the result array
- Tested the code well

zaelgohary reviewed (commented):
Now listing works fine as it lists the leader not the worker, however, on adding another worker after deploying, I faced this error.

![image](https://github.com/user-attachments/assets/df2c678a-eb56-4855-bc1d-45b640bcddd4)

Mahmoud-Emad commented:
> Now listing works fine as it lists the leader not the worker, however, on adding another worker after deploying, I faced this error.
> 
> ![image](https://private-user-images.githubusercontent.com/40770501/369830350-df2c678a-eb56-4855-bc1d-45b640bcddd4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjcxNjI2NDUsIm5iZiI6MTcyNzE2MjM0NSwicGF0aCI6Ii80MDc3MDUwMS8zNjk4MzAzNTAtZGYyYzY3OGEtZWI1Ni00ODU1LWJjMWQtNDViNjQwYmNkZGQ0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTI0VDA3MTkwNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTAwMzFlYWRmMjVlODc3MGMzNzJjNjUwZDI5NjcxN2NkZmM4OTZjNTA2MTUyZGMzYjViMzRjZDA0ZTQwYWE3MzgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.JwyM0if6vPacNE9Y8CRAjNADTZQB4DOSf_DKuYODgOU)

This issue isn't related to my PR, it could be another issue, it says there is another network workload with the same name already exists.

zaelgohary commented on packages/playground/src/weblets/tf_deployment_list.vue:
```diff
@@ -476,53 +486,112 @@ const deletingDialog = ref(false);
-        if (props.projectName?.toLowerCase() === ProjectName.Domains.toLowerCase()) {
-          await deleteGatewayDeployment(
-            updateGrid(grid, { projectName: props.projectName.toLocaleLowerCase() }),
-            item[0].workloads[0].name as string,
-          );
-        } else {
-          await deleteDeployment(updateGrid(grid!, { projectName: item.projectName }), {
-            deploymentName: item.deploymentName,
-            name: k8s ? item.deploymentName : item.name,
-            projectName: item.projectName,
-            ip: item.interfaces?.[0]?.ip,
-            k8s,
-          });
-        }
-      } catch (e: any) {
-        createCustomToast(`Failed to delete deployment with name: ${item.name}`, ToastType.danger);
-        console.log("Error while deleting deployment", e.message);
-        continue;
-      }
+    if (isDomainProject()) {
+      await deleteDomainProject(item);
+    } else {
+      await deleteStandardItem(item, k8s);
     }
-    table.value?.loadDeployments();
-  } catch (e) {
-    createCustomToast((e as Error).message, ToastType.danger);
-  } finally {
-    selectedItems.value = [];
-    deleting.value = false;
+  } catch (e: any) {
+    logDeletionError(item.name, e);
   }
 }
```
Do we need to separate these into 3 functions?

Mahmoud-Emad replied:
Why not?

maayarosama commented:
Why is most of the code separated into different function and only used once, for example ```prepareForDeletion``` and ```finalizeDeletion```. Is it because this's cleaner code somehow??

maayarosama commented:
I deployed a cluster and added two workers successfully. Listing of workers is fine
![Screenshot from 2024-09-24 13-15-56](https://github.com/user-attachments/assets/53625ae9-7ec8-4617-8312-929dc7bf07d7)
![Screenshot from 2024-09-24 13-16-21](https://github.com/user-attachments/assets/fda66c77-1aff-4765-9ed6-8307067dd08b)

However on shoe details, workers are not shown
![Screenshot from 2024-09-24 13-24-58](https://github.com/user-attachments/assets/47c9a2e0-45db-44a9-af8c-93d52faf69e1)

zaelgohary reviewed (changes requested):
I couldn't view domain deployment on this branch but it worked normally on staging.

![image](https://github.com/user-attachments/assets/f376401a-f32f-45a9-afdf-a13e557a1f01)

![image](https://github.com/user-attachments/assets/b4d967ba-6aa9-46c8-ba49-7ad25254b3a8)

Mahmoud-Emad committed:
Merge branch 'development' into development_caprover

Mahmoud-Emad commented:
@zaelgohary 


Working well with me, please make sure that you have the latest code and do  `yarn && make build`


https://github.com/user-attachments/assets/21131a58-1886-4db6-b3ec-89e1c52046e8

Mahmoud-Emad committed:
Merge branch 'development' of https://github.com/threefoldtech/tfgrid-sdk-ts into development_caprover

Mahmoud-Emad committed:
fix: Fix deleting Caprover cluster:

- Defined a new function to set the Caprover workers on the Caprover leader
- Delete the workers when delete the leader
- Remove not needed functions

zaelgohary commented on packages/playground/src/utils/deploy_helpers.ts:
```diff
@@ -11,6 +11,30 @@ export function createNetwork(network: Network = {}): NetworkModel {
   return nw;
 }
 
+export function setCaproverWorkers(vms: ZmachineData[], projectName: string | undefined = undefined): ZmachineData {
```
Why do we expect a projectName while we're setting caprover workers?

zaelgohary replied:
Won't the project in this case always be Caprover?

Mahmoud-Emad replied:
> Why do we expect a projectName while we're setting caprover workers?

if i understand well, we save the workers as the leader name to be able to load them, so why don't I hard-code it? I don't want to set it in all of the calls. I want to set it only after deploying a new worker

zaelgohary commented:
> Working well with me, please make sure that you have the latest code and do `yarn && make build`
> 
>  Screencast.from.2024-09-24.15-13-48.mp4

Can you click on show details as the error appears once clicking it?

I pulled latest changes and did yarn & make build.

![image](https://github.com/user-attachments/assets/3627b1d4-e799-4d95-98a4-30dc2c8d957d)

Mahmoud-Emad committed:
fix: Check if there is a project name before getting its data.

zaelgohary commented:
Workers are being showed as leaders in contracts page.

![image](https://github.com/user-attachments/assets/1c1f2603-ec42-4f8a-8159-f7731e517314)

![image](https://github.com/user-attachments/assets/0133bf3f-e36b-425f-aea2-fee5436d741e)



![image](https://github.com/user-attachments/assets/021dd262-b3bf-4f92-862b-f2be5f6a6ae2)

zaelgohary commented:
I deleted the caprover deployment but 1 network contract still exist after deletion. Happened more than once.

![image](https://github.com/user-attachments/assets/cec0053e-a44e-4887-a901-7848ee601636)

AhmedHanafy725 commented:
> Workers are being showed as leaders in contracts page.
> 
> ![image](https://private-user-images.githubusercontent.com/40770501/371094687-1c1f2603-ec42-4f8a-8159-f7731e517314.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjczNDczNzksIm5iZiI6MTcyNzM0NzA3OSwicGF0aCI6Ii80MDc3MDUwMS8zNzEwOTQ2ODctMWMxZjI2MDMtZWM0Mi00ZjhhLTgxNTktZjc3MzFlNTE3MzE0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTI2VDEwMzc1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdmMTU2MTFiZTZiNmU2ZmFjNDg0ODMzNDNhNDE0MTNjZmRhNTdiNzVlODM4OTAzNGRjY2M0OGU2MjBhNjI2NTgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.VTwTl-Q2TrIlbM0EE0mc5fcY02WeNiS_L2dxbf7Sshs)
> 
> ![image](https://private-user-images.githubusercontent.com/40770501/371094871-0133bf3f-e36b-425f-aea2-fee5436d741e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjczNDczNzksIm5iZiI6MTcyNzM0NzA3OSwicGF0aCI6Ii80MDc3MDUwMS8zNzEwOTQ4NzEtMDEzM2JmM2YtZTM2Yi00MjVmLWFlYTItZmVlNTQzNmQ3NDFlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTI2VDEwMzc1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWExYzAyMzNmMzk2YzFhMWMyZDA5NTlmZWRkNGY0MjMwZjQ3OWExMmZiZmI5YTlmZDYwODc4NTA2NDIyMjk2YmEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.S5iLsUbvfXybqgjxnFvdLA_zkkbCzr14QteLkpfEvLY)
> 
> ![image](https://private-user-images.githubusercontent.com/40770501/371094379-021dd262-b3bf-4f92-862b-f2be5f6a6ae2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjczNDczNzksIm5iZiI6MTcyNzM0NzA3OSwicGF0aCI6Ii80MDc3MDUwMS8zNzEwOTQzNzktMDIxZGQyNjItYjNiZi00ZjkyLTg2MmItZjJiZTVmNmE2YWUyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTI2VDEwMzc1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTkwMTMwZWRjODQ1M2RiZDI4Y2ZlZGU2NDMyYzZhOTUxMzA5NzNkYjkyNjMyY2I4YWU4ZjEyN2Q1NDYwMjdmNzQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.lasbvTtY7d-GRg8x-aBfaHEUm0Obs1-4SwAFGHNls5I)

@zaelgohary In the contracts, all the cluster's workers carry the same name as the cluster (which is the leader name) so we can list them as part of the cluster

Mahmoud-Emad commented:
> I deleted the caprover deployment but 1 network contract still exist after deletion. Happened more than once.
> 
> ![image](https://private-user-images.githubusercontent.com/40770501/371096025-cec0053e-a44e-4887-a901-7848ee601636.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjczNTAzNTcsIm5iZiI6MTcyNzM1MDA1NywicGF0aCI6Ii80MDc3MDUwMS8zNzEwOTYwMjUtY2VjMDA1M2UtYTQ0ZS00ODg3LWE5MDEtNzg0OGVlNjAxNjM2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTI2VDExMjczN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk1NWUwMTJmODRmMTA3YzcwNzg0NzFlN2YyOTQxMWY5N2M5ZTRjYWNiZjI2MWI2ZWJhOGRmYjI5ZTQyOTNkYzkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.v9VuGeDKTbEO5-QYCupWRHf-3UI73zGABVK8lirtrDU)

It's a known issue i think https://github.com/threefoldtech/tfgrid-sdk-ts/issues/3425

zaelgohary commented:
> > I deleted the caprover deployment but 1 network contract still exist after deletion. Happened more than once.
> > ![image](https://private-user-images.githubusercontent.com/40770501/371096025-cec0053e-a44e-4887-a901-7848ee601636.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjczNTAzNTcsIm5iZiI6MTcyNzM1MDA1NywicGF0aCI6Ii80MDc3MDUwMS8zNzEwOTYwMjUtY2VjMDA1M2UtYTQ0ZS00ODg3LWE5MDEtNzg0OGVlNjAxNjM2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTI2VDExMjczN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk1NWUwMTJmODRmMTA3YzcwNzg0NzFlN2YyOTQxMWY5N2M5ZTRjYWNiZjI2MWI2ZWJhOGRmYjI5ZTQyOTNkYzkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.v9VuGeDKTbEO5-QYCupWRHf-3UI73zGABVK8lirtrDU)
> 
> It's a known issue i think #3425

It's not. It was fixed in https://github.com/threefoldtech/tfgrid-sdk-ts/pull/3456 which was merged yesterday. Can you merge development to your branch? and if the issue still exist, we can open an issue for it as this only happen in Caprover.

Mahmoud-Emad commented:
The issue is already solved yeah, I just noticed that I have to build the client first 

![image](https://github.com/user-attachments/assets/b48b2b31-5949-4f59-84c6-51667bbbd39a)

However, yeah one contract still exists and should be deleted!

https://github.com/threefoldtech/tfgrid-sdk-ts/issues/3473

0oM4R commented on packages/playground/src/weblets/tf_deployment_list.vue:
```diff
@@ -484,29 +495,39 @@ watch(activeTab, () => (selectedItems.value = []));
+          }
+        } catch (e: any) {
+          createCustomToast(`Failed to delete deployment with name: ${item.name}`, ToastType.danger);
+          console.log("Error while deleting deployment", e.message);
```
please use console.error to make it clear in the logger
```suggestion
          console.error("Error while deleting deployment", e.message);
```

Mahmoud-Emad committed:
Merge branch 'development' into development_caprover

Mahmoud-Emad committed:
fix: Use console.error instead of console.log

zaelgohary reviewed (changes requested):
I tried listing a deployment I have but got the following errors. Am I doing anything wrong?

![image](https://github.com/user-attachments/assets/1447d395-f245-4937-90a3-8e75a307401f)

![image](https://github.com/user-attachments/assets/b3af9c04-e614-4e86-a739-23f08f0f3092)

zaelgohary commented:
I deployed a new instance & whenever I view its details I get the following error:

![image](https://github.com/user-attachments/assets/0a4f33c5-d7fc-427b-873b-14a607ddfaf4)

I also got the same error when deleting a worker.

maayarosama commented:
I tried adding workers most of the times it passed, but it failed once. Deletion works fine
![Screenshot from 2024-09-29 13-22-25](https://github.com/user-attachments/assets/c8f3f407-deb0-435a-94c3-817b33d82c62)
![Screenshot from 2024-09-29 13-24-07](https://github.com/user-attachments/assets/05ad2d33-ad80-4741-876f-f22abcbc28eb)

Mahmoud-Emad committed:
Merge branch 'development' of https://github.com/threefoldtech/tfgrid-sdk-ts into development_caprover

Mahmoud-Emad committed:
Merge branch 'development' into development_caprover

Mahmoud-Emad committed:
fix: Remove workers from the deployment list of the leader deleted

- Ensure that workers are correctly removed from the deployment list when the leader is deleted to maintain data integrity.
- Merged the latest changes from the development branch to keep the codebase up to date.
- Conducted thorough testing to verify the complete functionality and ensure no regressions were introduced.

Test Scenarios

1. Deploy `leader`, and `worker`, Go go to the contracts page and delete the deployed `leader`.
2. Deploy `leader`, attach a `worker` to it then delete the whole cluster just in time without refreshing the page.
3. Deploy `leader`, and `worker`, delete the deployed `worker` and attach another, then delete the whole cluster just in time without refreshing the page.
4. Deploy a cluster with 1 leader and 2 workers check the deployment listing dialog and make sure all of them are listed
3. Delete a cluster with the leader and more than 3 workers and make sure all of them are deleted.

PS: Make sure no errors on the Caprover listing page.

zaelgohary reviewed (approved):
Works fine adding/deleting clusters & works.

![image](https://github.com/user-attachments/assets/0f9b1dbf-1a77-4fd6-80b5-a380b42e0101)
